teuthology/orchestra/daemon/systemd.py

```diff
@@ -95,7 +95,11 @@ def parse_line(line):
     @property
     def pid(self):
         proc_name = 'ceph-%s' % self.type_
-        proc_regex = '"%s.*--id %s"' % (proc_name, self.id_)
+        proc_regex = '"%s.*--id %s "' % (proc_name, self.id_)
```

Comment:
I guess it would be better to have here more intelligent if-else coding, for example this regex is only make sense for `self.type_=='osd'`?
Also, the code is not working if the `--id` argument is the last argument in the command line?

Reply:
Also what do you think if we could parse ps output on client side with python regex so we can provide better test coverage?

Reply:
Hi @kshtsk, 

The regex which has `--id` is working fine and suitable for most of the daemons like (MON, OSD, MGR and MDS), since these daemon's has similar process command syntax as mentioned below.
But only radosgw needs different command regex as it has entirely different process command.

```
Rados-gw ps command and its proc regex:
=======================================
$ ssh -q magna111 "ps -ef | grep ceph"
ceph     13142     1  0 03:01 ?        00:01:55 /usr/bin/radosgw -f --cluster ceph --name client.rgw.magna111.rgw0 --setuser ceph --setgroup ceph

2020-04-13T14:59:55.507 INFO:teuthology.orchestra.run.magna111:> ps -ef | grep radosgw.*--name.*rgw.magna111 | grep -v grep | awk {'print $2'}
2020-04-13T14:59:55.572 INFO:teuthology.orchestra.run.magna111.stdout:13142

$ ps -ef | grep radosgw.*--name.*rgw.magna111 | grep -v grep | awk {'print $2'}
13142
```

for MON, OSD, MGR, MDS daemons, still same proc regex works fine. 
```
$ ps -ef | grep ceph
ceph      201807       1  0 02:52 ?        00:01:36 /usr/bin/ceph-osd -f --cluster ceph --id 6 --setuser ceph --setgroup ceph
ceph      203583       1  0 02:59 ?        00:00:25 /usr/bin/ceph-mds -f --cluster ceph --id magna047 --setuser ceph --setgroup ceph

$ ps -ef | grep ceph
ceph      101930       1  0 02:44 ?        00:02:01 /usr/bin/ceph-mon -f --cluster ceph --id magna017 --setuser ceph --setgroup ceph
ceph      103497       1  0 02:48 ?        00:03:59 /usr/bin/ceph-mgr -f --cluster ceph --id magna017 --setuser ceph --setgroup ceph
ceph      105283       1  0 02:57 ?        00:01:34 /usr/bin/ceph-osd -f --cluster ceph --id 10 --setuser ceph --setgroup ceph


2020-04-13T14:59:52.920 INFO:teuthology.orchestra.run.extensa019:> ps -ef | grep "ceph-mon.*--id extensa019 " | grep -v grep | awk {'print $2'}
2020-04-13T14:59:52.986 INFO:teuthology.orchestra.run.extensa019.stdout:19323

example:
$ ps -ef | grep "ceph-mon.* --id magna008 " | grep -v grep | awk {'print $2'}
28393


2020-04-06T03:16:48.976 INFO:teuthology.orchestra.run.extensa023:> ps -ef | grep "ceph-mds.*--id extensa023 " | grep -v grep | awk {'print $2'}
2020-04-06T03:16:49.042 INFO:teuthology.orchestra.run.extensa023.stdout:16749

example:
$ ps -ef | grep "ceph-mds.* --id magna047 " | grep -v grep | awk {'print $2'}
203583


2020-04-06T03:16:45.423 INFO:teuthology.orchestra.run.extensa025:> ps -ef | grep "ceph-osd.*--id 4 " | grep -v grep | awk {'print $2'}
2020-04-06T03:16:45.490 INFO:teuthology.orchestra.run.extensa025.stdout:14155

example:
$ ps -ef | grep "ceph-osd.* --id 1 " | grep -v grep | awk {'print $2'}
201606


2020-04-06T03:16:02.479 INFO:teuthology.orchestra.run.extensa022:> ps -ef | grep "ceph-mgr.*--id extensa022 " | grep -v grep | awk {'print $2'}
2020-04-06T03:16:02.548 INFO:teuthology.orchestra.run.extensa022.stdout:21451

example:
$ ps -ef | grep "ceph-mgr.* --id magna017 " | grep -v grep | awk {'print $2'}
103497


2020-04-13T14:59:55.507 INFO:teuthology.orchestra.run.extensa025:> ps -ef | grep radosgw.*--name.*rgw.extensa025 | grep -v grep | awk {'print $2'}
2020-04-13T14:59:55.572 INFO:teuthology.orchestra.run.extensa025.stdout:17213

example:
$ ps -ef | grep radosgw.*--name.*rgw.magna111 | grep -v grep | awk {'print $2'}
13142
```

Reply:
So what about adding a unit test?
the code can be refactored to make it possible, first, grab ps out like:
```
ps_out = self.remote.sh('ps -ef | grep -sE "/usr/bin/(ceph-osd|ceph-mds|ceph-mgr|ceph-mon|radosgw)"').strip()
```
And use regex in python on client side?

Reply:
I think this one is generally hard to unit test, you have to patch so much and the advantage is so less (and not really an infra code),  The test code is working fine and we could rely on that.